src/webview/html/messageAsHtml.js

```diff
@@ -72,14 +79,98 @@ $!${messageReactionListAsHtml(reactions, ownUser.user_id, allImageEmojiById)}
 `;
 };
 
-const widgetBody = (message: Message | Outbox) => template`
+/**
+ * Render the body of a message that has submessages.
+ *
+ * Must not be called on a message without any submessages.
+ */
+const widgetBody = (message: Message, ownUserId: UserId) => {
+  invariant(
+    message.submessages !== undefined && message.submessages.length > 0,
+    'should have submessages',
+  );
+
+  const widgetSubmessages: Array<{
+    sender_id: number,
+    content: SubmessageData,
+    ...
+  }> = message.submessages
+    .concat()
+    .sort((m1, m2) => m1.id - m2.id)
+    .filter(submessage => submessage.msg_type === 'widget')
```

Comment:
Not at all essential, but: if we put the `filter` before the `sort`, we can drop the `concat`, because the `filter` will already be creating a new array anyway.